File: deepspeed/inference/v2/kernels/ragged_ops/moe_gather/moe_gather.py
# Copyright (c) Microsoft Corporation.
# SPDX-License-Identifier: Apache-2.0

# DeepSpeed Team

import torch

from ... import DSKernelBase
from ....inference_utils import DtypeEnum
from deepspeed.ops.op_builder import RaggedOpsBuilder


class MoEGather(DSKernelBase):
    """
    CUDA implementation of MoE gather. This will bring the tokens back
    to their original indices and perform the output scaling.
    """

    supported_dtypes = [DtypeEnum.fp16, DtypeEnum.bf16]

    def __init__(self, dtype: DtypeEnum, channels: int, normalize_scores: bool = False) -> None:

        if not isinstance(dtype, DtypeEnum):
            dtype = DtypeEnum(dtype)

        if dtype not in MoEGather.supported_dtypes:
            raise RuntimeError(f"Unsupported dtype {dtype}")

        if channels % 8 != 0:
            raise RuntimeError(f"Channels {channels} must be divisible by 8")

        inf_module = RaggedOpsBuilder().load()
        self.kernel = inf_module.moe_gather
        self.normalize_scores = normalize_scores

    def __call__(self, layer_output: torch.Tensor, moe_output: torch.Tensor, scores: torch.Tensor,
                 mapped_slots: torch.Tensor, expert_counts: torch.Tensor) -> torch.Tensor:
        """
        Reorders the moe_output tokens into their original order and scales them by their
        gating scale. This will be a no-op for padded tokens.

        Arguments:
            layer_output (torch.Tensor): The output of the layer of shape [n_tokens, hidden_size]. This has been scaled appropriately.
            moe_output (torch.Tensor): The output of the MoE of shape [n_tokens * n_top_k, hidden_size].
            scores (torch.Tensor): The gating scores of shape [n_tokens].
            mapped_slots (torch.Tensor): The index of the token in the expert's input of shape [n_tokens, n_top_k]. The indices of token ``i`` in layer_output is ``mapped_slots[i]``.
            expert_counts (torch.Tensor): The number of tokens assigned to each expert of shape [n_experts]. This is passed to fuse the clearing of this data structure into the gather.

        Returns:
            layer_output
        """
        self.kernel(layer_output, moe_output, scores, mapped_slots, expert_counts, self.normalize_scores)
        return layer_output

File: deepspeed/inference/v2/kernels/ragged_ops/moe_gather/moe_gather.cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0

// DeepSpeed Team

#include "moe_gather.h"
#include <c10/cuda/CUDAStream.h>

#define DISPATCH_MOE_GATHER(T_TYPE, C_TYPE)                        \
    if (layer_output.options().dtype() == torch::T_TYPE) {         \
        launch_moe_gather((C_TYPE*)layer_output.data_ptr(),        \
                          (const C_TYPE*)moe_output.data_ptr(),    \
                          (const float*)scores.data_ptr(),         \
                          (const int32_t*)mapped_slots.data_ptr(), \
                          (int32_t*)expert_count.data_ptr(),       \
                          n_channels,                              \
                          n_experts,                               \
                          n_tokens,                                \
                          n_top_k,                                 \
                          normalize_scales,                        \
                          at::cuda::getCurrentCUDAStream());       \
        return;                                                    \
    }

/*
Re-gather the outputs of MoE and scale them by the gating score.
*/
void moe_gather(torch::Tensor& layer_output,
                const torch::Tensor& moe_output,
                const torch::Tensor& scores,
                const torch::Tensor& mapped_slots,
                const torch::Tensor& expert_count,
                const bool normalize_scales)
{
    const int32_t n_channels = layer_output.size(1);
    const int32_t n_experts = expert_count.size(0);
    const int32_t n_tokens = layer_output.size(0);
    const int32_t n_top_k = mapped_slots.size(1);

    TORCH_CHECK(moe_output.size(0) == n_tokens * n_top_k);
    TORCH_CHECK(moe_output.size(1) == n_channels);
    TORCH_CHECK(scores.size(0) == n_tokens);
    TORCH_CHECK(mapped_slots.size(0) == n_tokens);

    TORCH_CHECK(scores.size(1) == n_top_k);

    TORCH_CHECK(layer_output.scalar_type() == moe_output.scalar_type());
    TORCH_CHECK(scores.scalar_type() == torch::kFloat32);
    TORCH_CHECK(mapped_slots.scalar_type() == torch::kInt32);
    TORCH_CHECK(expert_count.scalar_type() == torch::kInt32);

    DISPATCH_MOE_GATHER(kHalf, __half);

#ifdef BF16_AVAILABLE
    DISPATCH_MOE_GATHER(kBFloat16, __nv_bfloat16);
#endif

    TORCH_CHECK(false, "Unsupported data type for MoE gather");
}
